Lib/fontTools/varLib/instancer.py

```diff
@@ -0,0 +1,1031 @@
+            continue
+        log.info("Instantiating FeatureVariations of %s table", tableTag)
+        _instantiateFeatureVariations(
+            varfont[tableTag].table, varfont["fvar"].axes, location
+        )
+        # remove unreferenced lookups
+        varfont[tableTag].prune_lookups()
+
+
+def _featureVariationRecordIsUnique(rec, seen):
+    conditionSet = []
+    for cond in rec.ConditionSet.ConditionTable:
+        if cond.Format != 1:
+            # can't tell whether this is duplicate, assume is unique
+            return True
+        conditionSet.append(
+            (cond.AxisIndex, cond.FilterRangeMinValue, cond.FilterRangeMaxValue)
+        )
+    # besides the set of conditions, we also include the FeatureTableSubstitution
+    # version to identify unique FeatureVariationRecords, even though only one
+    # version is currently defined. It's theoretically possible that multiple
+    # records with same conditions but different substitution table version be
+    # present in the same font for backward compatibility.
+    recordKey = frozenset([rec.FeatureTableSubstitution.Version] + conditionSet)
+    if recordKey in seen:
+        return False
+    else:
+        seen.add(recordKey)  # side effect
+        return True
+
+
+def _instantiateFeatureVariations(table, fvarAxes, location):
+    pinnedAxes = set(location.keys())
+    axisOrder = [axis.axisTag for axis in fvarAxes if axis.axisTag not in pinnedAxes]
+    axisIndexMap = {axisTag: axisOrder.index(axisTag) for axisTag in axisOrder}
+
+    featureVariationApplied = False
+    uniqueRecords = set()
+    newRecords = []
+
+    for i, record in enumerate(table.FeatureVariations.FeatureVariationRecord):
+        retainRecord = True
+        applies = True
+        newConditions = []
+        for j, condition in enumerate(record.ConditionSet.ConditionTable):
+            if condition.Format == 1:
+                axisIdx = condition.AxisIndex
+                axisTag = fvarAxes[axisIdx].axisTag
+                if axisTag in pinnedAxes:
+                    minValue = condition.FilterRangeMinValue
+                    maxValue = condition.FilterRangeMaxValue
+                    v = location[axisTag]
+                    if not (minValue <= v <= maxValue):
+                        # condition not met so remove entire record
+                        retainRecord = applies = False
+                        break
+                else:
+                    # axis not pinned, keep condition with remapped axis index
+                    applies = False
+                    condition.AxisIndex = axisIndexMap[axisTag]
+                    newConditions.append(condition)
+            else:
+                log.warning(
+                    "Condition table {0} of FeatureVariationRecord {1} has "
+                    "unsupported format ({2}); ignored".format(j, i, condition.Format)
+                )
+                applies = False
+                newConditions.append(condition)
```

Comment:
suggest rewrite to not have the deep nesting, either reordering and/or using narrow purpose functions.

For example [very pseudocode]:

```
# _instantiateFeatureVariations
for i, record in ...
  instantiateFeatureVariation

# instantiateFeatureVariation
# loop and call fn for every condition

# fn for every condition
if format != 1
  log
  return False
if axisTag not in pinnedAxes
  ...
  return False
...now do stuff...
```